folly/CPortability.cpp

```diff
@@ -0,0 +1,563 @@
+  ZeroMemory(s, n);
+}
+
+int closedir(DIR* dir) {
+  if (!FindClose(dir->hand))
+    return -1;
+  free(dir);
+  return 0;
+}
+
+char* ctime_r(const time_t* t, char* buf) {
+  char tmpBuf[64];
+  if (ctime_s(tmpBuf, sizeof(tmpBuf), t))
+    return NULL;
+  return strcpy(buf, tmpBuf);
+}
+
+char* dirname(char* path) {
+  if (path == NULL || !strcmp(path, ""))
+    return ".";
+
+  size_t len = strlen(path);
+  char* pos = strrchr(path, '/');
+  if (strrchr(path, '\\') > pos)
+    pos = strrchr(path, '\\');
+  if (pos == NULL)
+    return ".";
+
+  // Final slash with no name.
+  if (path + len == pos) {
+    *pos = '\0';
+
+    char* pos2 = strrchr(path, '/');
+    if (strrchr(path, '\\') > pos2)
+      pos2 = strrchr(path, '\\');
+    if (pos2 == NULL)
+      return ".";
+
+    *pos2 = '\0';
+    return path;
+  }
+
+  *pos = '\0';
+  return path;
+}
+
+int dprintf(
+  int fd,
+  _Printf_format_string_ const char *fmt,
+  ...
+  ) {
+  va_list args;
+  va_start(args, fmt);
+
+  int len = vsnprintf(NULL, 0, fmt, args);
+  if (len <= 0)
+    return -1;
+  char* buf = (char*)malloc(len + 1);
+  if (vsnprintf(buf, len + 1, fmt, args) == len &&
+    _write(fd, buf, len) == len) {
+    free(buf);
+    va_end(args);
+    return len;
+  }
+
+  free(buf);
+  va_end(args);
+  return -1;
+}
+
+int finite(double d) {
+  return isfinite(d) ? 1 : 0;
+}
+
+int getgid() {
+  return 1;
+}
+
+// No major need to implement this, and getting a non-potentially
+// stale ID on windows is a bit involved.
+pid_t getppid() {
+  return (pid_t)1;
+}
+
+int getrlimit(int type, rlimit* dst) {
+  if (type == RLIMIT_STACK) {
+    NT_TIB* tib = (NT_TIB*)NtCurrentTeb();
+    dst->rlim_max = dst->rlim_cur = (size_t)tib->StackBase - (size_t)tib->StackLimit;
+    return 0;
+  }
+  return -1;
+}
+
+int getrusage(int who, rusage* usage) {
+  ZeroMemory(usage, sizeof(rusage));
+  return 0;
+}
+
+int gettimeofday(timeval* tv, timezone*) {
+  FILETIME ft;
+  if (tv) {
+    GetSystemTimeAsFileTime(&ft);
+    uint64_t ns = *(uint64_t*)&ft;
+    tv->tv_usec = (long)((ns / 10ULL) % 1000000ULL);
+    tv->tv_sec = (long)((ns - POSIX_WIN_FT_OFFSET) / 10000000ULL);
+  }
+  return 0;
+}
+
+int getuid() {
+  return 1;
+}
+
+tm* gmtime_r(const time_t* t, tm* res) {
+  if (!gmtime_s(res, t))
+    return res;
+  return NULL;
+}
+
+int kill(pid_t p, int sig) {
+  HANDLE h = OpenProcess(PROCESS_TERMINATE, FALSE, (DWORD)p);
+  if (!TerminateProcess(h, (UINT)-1))
+    return -1;
+  return 0;
+}
+
+tm* localtime_r(const time_t* t, tm* o) {
+  tm* tmp = localtime(t);
+  if (tmp) {
+    *o = *tmp;
+    return o;
+  }
+  return NULL;
+}
+
+int madvise(const void* addr, size_t len, int advise) {
+  // We do nothing at all.
+  // Could probably implement dontneed via VirtualAlloc
+  // with the MEM_RESET and MEM_RESET_UNDO flags. 
+  return 0;
+}
+
+size_t malloc_usable_size(void* addr) {
+  return _msize(addr);
+}
+
+void* memmem(const void* haystack, size_t hlen, const void* needle, size_t nlen) {
```

Comment:
Since this is Windows-only, perhaps you could annotate this function like so:

```
_Ret_maybenull_ memmem( _In_reads_bytes_( hlen) const void* haystack, size_t hlen, _In_reads_bytes_( nlen ) const void* needle, size_t nlen )
```